Stb image loading

marauder2k7 opened the pull request:
Mark this for 4.2

NOTES: Screenshot needs updated to use stb, was originally able to tile for larger than normal res screenshots need to figure ot if this is possible in stb image or if we need to just do it in gbitmap.

blTerrainSystem lightmap needs to be changed to use a filename, or if we could figure out how to marry Stream and the stbimage loaders we could revert back from using a filepath and use stream instead.

marauder2k7 committed:
STBImage

STB used for loading and writing images.

marauder2k7 committed:
Merge remote-tracking branch 'upstream/development' into STBImageLoading-PR

marauder2k7 committed:
Bug fixes:

Generating image previews of image assets was failing

DDS remove redundant check for stream status.

STB requires the file to be free before being written to, move check to make sure we can open the path into gBitmap and remove FileStream checks from everywhere else.

dottools commented on Engine/source/gfx/screenshot.cpp:
```diff
@@ -130,85 +130,85 @@ void ScreenShot::capture( GuiCanvas *canvas )
+   //            U8 *blendOverlapSrc = (U8*)inColor;
+   //            U8 *blendOverlapDst = outColor - inOverlapOffset;
+   //            for ( U32 px=0; px < overlapPixels.x; px++)
+   //            {
+   //               F32 blendFactor = (F32)px / (F32)overlapPixels.x;
+   //               sBlendPixelRGB888(blendOverlapSrc, blendOverlapDst, blendFactor);                 
+
+   //               blendOverlapSrc += gb->getBytesPerPixel();
+   //               blendOverlapDst += outBuffer->getBytesPerPixel();                   
+   //            }               
+   //         }
+
+   //         //Gradient blend against the rows the excess overlap rows already in the buffer            
+   //         if (ty && row < overlapPixels.y)
+   //         {
+   //            F32 rowBlendFactor = (F32)row / (F32)overlapPixels.y;
+   //            U8 *blendSrc = outColor + outStride * (outBuffer->getHeight() - overlapPixels.y);
+   //            U8 *blendDst = outColor;               
+   //            for ( U32 px=0; px < gb->getWidth() - overlapPixels.x*2; px++)
+   //            {                  
+   //               sBlendPixelRGB888(blendSrc, blendDst, 1.0-rowBlendFactor); 
+   //               blendSrc += gb->getBytesPerPixel();
+   //               blendDst += outBuffer->getBytesPerPixel();                   
+   //            }                              
+   //         }
+
+   //         
+   //         inColor += inStride;
+   //         outColor += outStride;
+   //      }
+
+   //      delete gb;
+   //   }
+
+   //   // Write the captured tile row into the PNG stream
+   //   pngWriter.append(outBuffer, outBuffer->getHeight()-overlapPixels.y);
+   //}
+
+   ////Close the PNG stream
+   //pngWriter.end();
```
Please learn to:
```cpp
#if 0
// code to disable goes here
#endif // 0
```

Instead of multi-line commenting out block of code to avoid these unnecessary large diffs.

marauder2k7 replied:
> Please learn to:
> 
> ```c++
> #if 0
> // code to disable goes here
> #endif // 0
> ```
> 
> Instead of multi-line commenting out block of code to avoid these unnecessary large diffs.

was talked over with az and jeff, this is just commented out for now to be changed when we figure out how to do this with stb image, essentially it is mostly going to be removed either way

dottools replied:
I understand, but for languages `C`, `C++`, and even `C#` multiline commenting out is just not acceptable for revision control when you can `#if 0` it out to disable code for the time being until it is actually revised or actually deleted.

marauder2k7 replied:
sorted

dottools commented on Engine/source/gfx/bitmap/loaders/bitmapSTB.cpp:
```diff
@@ -0,0 +1,226 @@
+
+   if (ext.equal("jpg") || ext.equal("jpeg"))
+   {
+      if (stbi_write_jpg(path.getFullPath().c_str(), width, height, comp, bitmap->getWritableBits(), 90))
```
Quality level should be a configurable value not hard coded to 90.

marauder2k7 replied:
this one i was thinking of using the compressionLevel argument in the function, but people wold need to know that it is a percentage so needs to be properly documented

dottools replied:
Ah ok. Here's an idea for documenting this function:
```cpp
/**
 * Write bitmap to an image file.
 *
 * @param[in]       path                        Destination image file path. File name extension determines image format.
 *                                              ".bmp" for Microsoft Bitmap (BMP).
 *                                              ".hdr" for High Dynamic Range (HDR).
 *                                              ".jpg" or ".jpeg" for Joint Photographic Experts Group (JPEG).
 *                                              ".png" for Portable Network Graphics (PNG).
 *                                              ".tga" for Truevision TGA (TARGA).
 *
 *
 * @param[in]       bitmap                      Source bitmap to encode image from.
 * @param           compressionLevel            Image format specific compression level.
 *                                              For JPEG sets the quality level percentage, range 0 to 100.
 *                                              Not used for other image formats.
 */
bool sWriteSTB(const Torque::Path& path, GBitmap* bitmap, U32 compressionLevel)
```

marauder2k7 replied:
nice thanks :) ill put that in, and ill add the documentation to gbitmap as well since it comes directly from that

marauder2k7 replied:
committed

marauder2k7 committed:
Update screenshot.cpp

Screenshot now works with stb: gBitmap outbuffer holds the whole image anyway so just use writeBitmap when its done.

marauder2k7 committed:
Added documentation

CompressionLevel argument now used to set jpeg quality as well. Values need to be 0-100 range for jpeg and 0-10 for png.

marauder2k7 committed:
Update gBitmap.h

typo

marauder2k7 committed:
Added hdr and bypass 16bit when png.

Added support for .hdr format, atm just being converted to ldr after the exposure is applied.

cleanup dds file

PNG skips 16bit altogether at the moment for textureArray issues.

marauder2k7 committed:
Merge remote-tracking branch 'upstream/development' into STBImageLoading-PR

marauder2k7 committed:
MacOSX fixes

Azaezel commented:
need to resolve an application close flaw, but otherwise seems good

marauder2k7 committed:
Merge remote-tracking branch 'upstream/development' into STBImageLoading-PR

marauder2k7 commented:
> need to resolve an application close flaw, but otherwise seems good

no longer able to trip this error after updating the pr to latest

marauder2k7 committed:
Update assetImport.tscript

add extra formats for stb_image to asset importer
